File: 3074.apple-redistribution-into-boxes.cpp
/*
 * @lc app=leetcode id=3074 lang=cpp
 *
 * [3074] Apple Redistribution into Boxes
 */

// @lc code=start
#include <vector>
#include <algorithm>

class Solution {
public:
    int minimumBoxes(std::vector<int>& apple, std::vector<int>& capacity) {
        std::sort(capacity.begin(), capacity.end(), std::greater<int>());
        std::vector<int> prefix = {0};
        for (int c : capacity) {
            prefix.push_back(prefix.back() + c);
        }
        int appleSum = 0;
        for (int a : apple) {
            appleSum += a;
        }
        return std::lower_bound(prefix.begin(), prefix.end(), appleSum) - prefix.begin();
    }
};

// @lc code=end

File: 3074.apple-redistribution-into-boxes.py
#
# @lc app=leetcode id=3074 lang=python3
#
# [3074] Apple Redistribution into Boxes
#

# @lc code=start
import bisect
from typing import List


class Solution:
    '''
    sort capacity from large to small, prepare prefix sum
    then one binary search for sum(apple)
    '''
    def minimumBoxes(self, apple: List[int], capacity: List[int]) -> int:
        capacity.sort(reverse=True)
        prefix = [0]
        for c in capacity: prefix.append(prefix[-1] + c)
        return bisect.bisect_left(prefix, sum(apple))
# @lc code=end

